staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go

```diff
@@ -290,17 +293,27 @@ func (a *GenericAdmissionWebhook) hookClient(h *v1alpha1.ExternalAdmissionHook)
 		return nil, err
 	}
 
-	// TODO: cache these instead of constructing one each time
-	restConfig, err := a.authInfoResolver.ClientConfigFor(serverName)
-	if err != nil {
-		return nil, err
+	a.mu.RLock()
+	restConfig, ok := a.restConfigCache[serverName]
+	a.mu.RUnlock()
+	if !ok {
+		restConfig, err = a.authInfoResolver.ClientConfigFor(serverName)
+		if err != nil {
+			return nil, err
+		}
+
+		a.mu.Lock()
+		a.restConfigCache[serverName] = restConfig
+		a.mu.Unlock()
 	}
+
```

Comment:
I'd go a step further and cache the constructed client based on the data in the ExternalAdmissionHook, if possible. Means the service resolution would need to be moved into the dialer so the same client would continue to work as the resolved IP changes

Reply:
I know your meaning, but the h *v1alpha1.ExternalAdmissionHook may change.

Reply:
Unless use `ExternalAdmissionHook` as key.

Reply:
Right, the relevant data in the admission config is the key

Reply:
@liggitt  I think I can cache client  later.

Reply:
client construction on every request is expensive... we'll need to revisit this if we don't do it now

Reply:
@liggitt cache client done